octobot_commons/databases/cache_manager.py

```diff
@@ -0,0 +1,205 @@
+                            self.__class__.CACHES[_tentacle_name][_exchange_name][_symbol]):
+                        yield self.__class__.CACHES[_tentacle_name][_exchange_name][_symbol][_time_frame], \
+                              (_tentacle_name, _exchange_name, _symbol, _time_frame)
+
+    def _open_or_create_cache_database(self, tentacle, exchange, symbol, time_frame,
+                                       tentacle_name, tentacles_setup_config, cache_type):
+        cache_full_path = self.get_cache_path(tentacle, exchange, symbol, time_frame,
+                                              tentacle_name, tentacles_setup_config)
+        cache_dir = os.path.split(cache_full_path)[0]
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
+        return self._open_cache_database(cache_full_path, cache_type)
+
+    def _open_cache_database(self, file_path, cache_type):
+        """
+        Override to use another cache database or adaptor
+        :return: the cache database class
+        """
+        return _CacheWrapper(file_path, cache_type, self.database_adaptor)
+
+    def get_cache_path(self, tentacle, exchange, symbol, time_frame, tentacle_name, tentacles_setup_config):
+        try:
+            return self.__class__.CACHES[tentacle_name][exchange][symbol][time_frame].get_path()
+        except KeyError:
+            sanitized_pair = symbol_util.merge_symbol(symbol) if symbol else symbol
+            # warning: very slow, should be called as rarely as possible
+            required_tentacles = self._get_linked_tentacles(tentacle, tentacles_setup_config)
+            identifying_tentacles = [tentacle] + required_tentacles
+            return os.path.join(common_constants.USER_FOLDER, common_constants.CACHE_FOLDER, tentacle_name,
+                                exchange, sanitized_pair, time_frame,
+                                self._code_hash(identifying_tentacles),
+                                self._config_hash(identifying_tentacles, tentacles_setup_config),
+                                common_constants.CACHE_FILE)
+
+    def _get_linked_tentacles(self, tentacle, tentacles_setup_config):
+        try:
+            import octobot_tentacles_manager.api as tentacles_manager_api
+            linked_tentacles_classes = tentacles_manager_api.get_tentacle_classes_requirements(tentacle)
+            tentacles = []
+            for tentacle_class in linked_tentacles_classes:
+                # try instantiating tentacles, if not possible, then ignore it. Works with evaluator tentacles.
+                try:
+                    tentacles.append(tentacle_class(tentacles_setup_config))
+                except TypeError:
+                    pass
+            return tentacles
+        except ImportError as e:
+            raise ImportError("octobot_tentacles_manager is required to use cache") from e
+
+    def _code_hash(self, identifying_tentacles) -> str:
```

Comment:
I think this code should be placed in OctoBot-Tentacles-Manager

Reply:
yes!